Show netgear devices without a link_rate as not_home

dlashua opened the pull request:
## Description:
Netgear device tracker takes 300+ seconds to see a device as no longer on the network. This PR fixes that by looking at the link_rate returned by the API. 

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** 

Documentation remains the same.

## Example entry for `configuration.yaml` (if applicable):
Config remains the same.

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

dlashua committed:
show devices without a link_rate as not_home

balloob commented on homeassistant/components/netgear/device_tracker.py:
```diff
@@ -110,6 +110,7 @@ def scan_devices(self):
                     or dev.name in self.excluded_devices
                 )
             )
+            tracked = tracked and dev.link_rate is not None
```
`tracked` keeps track if a device is tracked. `link_rate` should not impact the value of this variable.

dlashua replied:
In this case, "tracked" is a boolean that determines if the device should be reported as seen or not. self.last_results contains all the seen devices. However, pynetgear returns devices even after they've disconnected from WiFi, however, for these disconnected devices, link_rate is None. 

Unless you just mean that you want link_rate to not contaminate the value of tracked, in which case, I can change it to a simple "if dev.link_rate is None" inside of the "if tracked:" block.

balloob replied:
Yeah, we should not contaminate the value. Also add a comment why `link_rate is None` is checked.

dlashua replied:
I have made the requested changes.

dlashua committed:
add comments, change style

MartinHjelmare commented on homeassistant/components/netgear/device_tracker.py:
```diff
@@ -111,12 +111,15 @@ def scan_devices(self):
-                    devices.append(dev.mac + "_" + dev.conn_ap_mac)
+                # if link_rate is None, the device is still remembered by the router however,
+                # it is no longer connected. Therefore, we will not report it as seen.
+                if dev.link_rate is not None:
```
This check can be combined with the check above, without changing the value of `tracked`. Just combine them with `and`.

fabaff committed:
Upgrade zeroconf to 0.24.3 (#30187)

Quentame committed:
Fix AdGuard Home safe search sensor name (#30171)

* Fix AdGuard Home safe search sensor name

* Fix typo

garbled1 committed:
Add onewire devices and owserver remote host support (#29948)

* Add support for HobbyBoards sensors: Hub, Humidity, Moisture, PulseCounter
Add support for an owserver running on a remote host

* Cleanup and fixes for style/etc

* Forgot to convert to f-strings.

* Update with changes requested by @MartinHjelmare

* Pylint

* Fix to None

* Reverse logic for hobbyboard test to eliminate pylint complaint

* Modified manifest to list myself as codeowner

* Move some functions down into OneWireProxy instead of the top OneWire

* Oops, missed some decode()'s

* And another!

springstan committed:
Bump importlib-metadata to 1.3.0 (#30196)

* Bump importlib-metadata to 1.3.0

* Include setup.py and package_constraints.txt

springstan committed:
Bump datapoint to 0.9.5 (#30185)

ochlocracy committed:
Implement EqualizerController in Alexa for media_player. (#30159)

homeassistant committed:
[ci skip] Translation update

BKPepe committed:
Upgrade youtube_dl to 2019.12.25 (#30203)

ndonegan committed:
Pull track position from MPD status (#28579)

* Pull track position from MPD status()

This allows the progress bar to work when using the media-control card in lovelace.

* Actually commit flake8 fix!

* Extra documentation.

Mainly to trigger CI rerun.

* Updated to use self._media_position

rhadamantys committed:
Add support for enocean window handle FA 10 00 (Hoppe) (#29968)

* Added support for Somfy RTS wireless power socket and
Somfy Temperature Sensore Thermos Wirefree io

* Added code formatting fixes for commit 5faaf9c

* added support for RollerShutterRTSComponent from Somfy

* Added support for RTS roller shutter in set_cover_position

* added support for enocean window handle FA 10 00 (Hoppe)

* removed spaces in empty lines

* removal of rawdata attribute / code style

* isort fix

* PyLint fixes

* Improvements after review suggestions: rename device class to sensor type to avoid ambiguousness

* added import for DEVICE_CLASS_POWER from const.py

* removed window handle unit

HarlemSquirrel committed:
Update Vivotek camera component (#30191)

- Bump libpyvivotek version to 0.4.0
- Add digest authentication

scop committed:
Make Huawei LTE notify service name configurable (#30208)

* Default Huawei LTE notify service name to notify.huawei_lte, make configurable

Closes https://github.com/home-assistant/home-assistant/issues/29409

* Set default notify service name for uninvoked options flow

scop committed:
Add Huawei LTE integration suspend and resume services (#30207)

Useful e.g. if accessing the router web interface from another source
such as a web browser is temporarily required.

homeassistant committed:
[ci skip] Translation update

alandtse committed:
Bump teslajsonpy to 0.2.1 (#30217)

Closes #29922

PaulAnnekov committed:
bump tuyaha 0.0.5 (#30213)

bendavid committed:
Fix handling of symlinked device descriptors in keyboard_remote and move remaining sync io to executor thread pool (#30206)

* fix handling of symlinked device decriptors

* make check for symlinked paths more efficient

* make variable names pylint compliant

* move sync io during setup and device connect/disconnect to executor thread pool

* move remaining sync io during setup to executor thread pool

* remove unnecessary lambda functions

dmulcahey committed:
Bump ZHA dependencies (#30228)

* bump dependencies
* requirement files
* bump quirks to 0.0.30

homeassistant committed:
[ci skip] Translation update

frenck committed:
Bump shodan to 1.21.1 (#30234)

Kerwood committed:
DECONZ - Added support for Aqara single switch WXKG03LM (#30240)

homeassistant committed:
[ci skip] Translation update

springstan committed:
Bump python-qbittorrent to 0.4.1 (#30239)

tchellomello committed:
Bump ring to 0.2.8 to fix Oauth issues (#30245)

scop committed:
Fix Huawei LTE error message on service call without URL and routers (#30250)

Commit:
Bump dependency for HomematicIp cloud (#30237)

* Bump dependency for HomematicIp cloud

* Update test_data

xlfe committed:
Bump libpurecool to 0.6.0 (#30247)

* bump libpurecool to 0.6.0 - fixes home-assistant/home-assistant#26367

* update manifest too

scop committed:
Huawei LTE: Fix YAML options overriding ones set from GUI (#30249)

Closes https://github.com/home-assistant/home-assistant/issues/30221

Commit:
Add HmIP-MOD_TM to HomematicIP Cloud (#30255)

scop committed:
Note Huawei LTE notify service change requires restart (#30223)

Refs https://github.com/home-assistant/home-assistant/issues/30222

cereal2nd committed:
Bump python-velbus library to 2.0.32 to fix problems with the glaspanels (#30257)

balloob committed:
Whitelist Android/iOS auth callbacks (#30082)

* Whitelist Android/iOS

* Add iOS alternate flavor URLs

* Update indieauth.py

Co-authored-by: Robbie Trencheny <me@robbiet.us>

vangorra committed:
Fixing timezone issue which caused wrong selection of data to be used. (#30011)

Adminiuga committed:
Refactor ZHA channel logging (#30259)

Add channel.id property -- id unique for this the device only.

homeassistant committed:
[ci skip] Translation update

liollury committed:
Fix ble_tracker randomly pygatt thrown error (#28671)

* fix(ble_tracker): catch randomly pygatt thrown error

* fix(ble_tracker): merge except errors

mockersf committed:
Fix creating smappee sensors when remote is not active (#30270)

scop committed:
Upgrade rflink to 0.0.50, ignore_devices now supports * and ? anywhere (#30268)

https://github.com/aequitas/python-rflink/releases/tag/0.0.50
https://github.com/aequitas/python-rflink/releases/tag/0.0.49
https://github.com/aequitas/python-rflink/releases/tag/0.0.48
https://github.com/aequitas/python-rflink/releases/tag/0.0.47

homeassistant committed:
[ci skip] Translation update

Jc2k committed:
Accept homekit_controller pairing codes both with and without dashes (#30273)

* Handle MalformedPinError from homekit_python
* Handle both formats of pin codes

Jc2k committed:
Add homekit_controller service.sensor.smoke (#30269)

fabaff committed:
Upgrade beautifulsoup4 to 4.8.2 (#30274)

fabaff committed:
Add custom validator for countries (#30280)

alandtse committed:
Change default icons for Tesla components (#30288)

andygrunwald committed:
Contributing: Add note about feature suggestions and bug tracking (#30225)

taisholland committed:
Bump ihcsdk to 2.4.0 (#30279)

* Update requirements_all.txt

update ihcsdk to version 2.4.0

* Update manifest.json

upgrade to version 2.4.0 of ihcsdk

Kane610 committed:
Handle wired bug on restart (#30276)

andrewsayre committed:
Bump pysmartthings 0.7.0 (#30302)

homeassistant committed:
[ci skip] Translation update

pnbruckner committed:
Ignore google_maps updates when last_seen goes backwards (#30178)

bieniu committed:
Add GIOS integration (#28719)

* Initial commit

* Add gios to requirements

* Add tests

* Update .coveragerc file

* Run gen_requirements_all.py

* Change DEFAULT_SCAN_INTERVAL

* Better strings

* Bump library version

* run script.hassfest

* run isort

* Add icons mapping

* Remove unnecessary f-string

* Remove unnecessary listener

* Refactoring config_flow

* Add unique_id to config entry

* Change AQI states to consts in English

* Remove unused init

* Remove unused exception

* Remove private instance attribute

* Remove overwrite state property

* Fix pylint error

* Add SCAN_INTERVAL for air_quality entity

* Add _abort_if_unique_id_configured()

balloob committed:
Store capabilities and supported features in entity registry, restore registered entities on startup (#30094)

* Store capabilities and supported features in entity registry

* Restore states at startup

* Restore non-disabled entities on HA start

* Fix test

* Pass device class from entity platform

* Clean up restored entities from state machine

* Fix Z-Wave test?

frenck committed:
Fix Withings leaking time zone change into other tests (#30320)

* Fix Withings leaking time zone change in other tests

* Fix spelling error in code doc

issacg committed:
Add local_ip component (#29973)

* Added localip component

* Split config and core logic, and migrate to sensor platform (requested by @MartinHjelmare)
Also allow overriding the sensor name via the config

* Tweak docstring

Co-Authored-By: Fabian Affolter <mail@fabian-affolter.ch>

* Initial support for config entries

* Rename localip to local_ip (1/2)

* Rename localip to local_ip (2/2)

* Add test for config_flow

* Split and rename tests

* Remove unneeded code from config_flow

* Implement configuration as config entry import.  Other misc requested changes from code review.

* Fix tests

* minor code review fixes

* remove unneeded code

Co-authored-by: Fabian Affolter <mail@fabian-affolter.ch>

pschmitt committed:
Update liveboxplay and pyteleloisirs (#30093)

sbyx committed:
Add homematic host port config for HMIP-only CCUs (#30077)

* homematic: Add host port config for HMIP-only CCUs

When adding a host (CCU) to the homematic component currently the hardcoded port 2001 is used to communicate with it.
However that port is only available on the target if the target supports HM (wireless) protocol which is not the case e.g. for
the Hass.io Homematic CCU addon when running in HMIP-only mode with the HMIP-RFUSB stick.

This allows to change the port home assistant uses to talk to the CCU in order to provide services under hte homematic homain, e.g. homematic.set_variable_value

The default value for this option is the old hardcoded value this the change should be backwards compatible with existing configurations.

* Change style of config retrieval

Commit:
Bump dependency for HomematicIP Cloud (#30319)

rhadamantys committed:
Add support for Somfy RTS power socket and Somfy io Temperature sensor (#30053)

* Added support for Somfy RTS wireless power socket and
Somfy Temperature Sensore Thermos Wirefree io

* Added code formatting fixes for commit 5faaf9c

* added support for RollerShutterRTSComponent from Somfy

* Added support for RTS roller shutter in set_cover_position

* Add support for Somfy RTS power socket and Somfy io temperature sensor

* black and isort fixes

sdwilsh committed:
Reduce solaredge logging severity (#30305)

* [solaredge] Reduce Severity of Log Line

This log error happens frequently for some sites, but it shouldn't be an error.
It is expected, per the SolarEdge Monitoring API, that some sites do not
support this information, and the expected result is that this would be
empty (see comments on #27959).

Fixes #27959

* Fix a typo

Co-authored-by: Fabian Affolter <mail@fabian-affolter.ch>

pnbruckner committed:
Remove Amcrest deprecated sensors and switches (#30308)

cwildt committed:
Add support for Velux garage doors (#30214)

* Update manifest.json

* Update cover.py

Add GarageDoor

* Update to pyvlx 0.2.12

* Sort

tuxbox committed:
Add separate command and state topics for mqtt lock (#29808)

* Update lock.py

Allow different command and state topic + different command and state values.

* Formatting updated after black run

* TC updated to reflect different state & cmd values

* Abbreviations for lock states added

* additional non-default state test

* whitespaces fixed

* black formatting run

fabaff committed:
Upgrade Sphinx to 2.3.1 (#30310)

brefra committed:
Add light support to Velbus integration (#30323)

* Add light support to Velbus integration

* Add Velbus light.py to .coveragerc

* Applied black formatting

Lars-P committed:
Fix luftdaten integration by adding a sensor for pressure at sealevel (#30317)

alandtse committed:
Simplify Tesla icon get code (#30301)

* Simplify icon get code

* Remove extraneous None

Adminiuga committed:
Make the rest of ZHA platforms to use ZHA class registry (#30261)

* Refactor ZHA component tests fixtures.

* Add tests for ZHA device discovery.

* Refactor ZHA registry MatchRule.

Allow callables as a matching criteria.
Allow sets for model & manufacturer.

* Minor ZHA class registry refactoring.

Less cluttered strict_matching registrations.

* Add entities only if there are any.

* Migrate rest of ZHA platforms to ZHA registry.

* Pylint fixes.

frenck committed:
Improve Withings tests in different time zone (#30326)

* Improve Withings tests in different time zone

* Address code review comment

* Spelling error in code doc

frenck committed:
Migrate dsmr tests from coroutine to async/await (#30333)

frenck committed:
Migrate wunderground tests from coroutine to async/await (#30330)

frenck committed:
Bump pyps4-2ndscreen to 1.0.4 (#30327)

springstan committed:
Bump alarmdecoder to 1.13.9 (#30303)

frenck committed:
Migrate python_script tests from coroutine to async/await (#30335)

frenck committed:
Migrate api tests from coroutine to async/await (#30338)

frenck committed:
Migrate alexa tests from coroutine to async/await (#30332)

frenck committed:
Migrate xiaomi_miio tests from coroutine to async/await (#30329)

frenck committed:
Migrate emulated_hue tests from coroutine to async/await (#30331)

bendavid committed:
Migrate webostv to new library and make integration async with callback state updates (#29296)

* migrate webostv to new aiopylgtv version of the library and add support
for generic commands, input/button commands, and callback state updates

* update requirements

* cleanup and bump aiopylgtv version

* update webostv unit tests

* make webostv unit tests work with python 3.7

* cleanup for code checks

* cleanup and code review

* make all client request functions coroutines

* make host required for webostv configuration

* remove generic command and button functionality plus related cleanup

* fix previous track function

* update unit tests

* fix imports for unit tests

* update unit test

* further unit test updates

* remove unnecessary setup call in unit tests

* restore previous behaviour with client key config file in hass configuration directory

frenck committed:
Migrate input_* tests from coroutine to async/await (#30336)

homeassistant committed:
[ci skip] Translation update

alandtse committed:
Remove use of bin_type in Tesla component (#30315)

* Remove use of bin_type

* Convert _unit attribute to units

* Remove unnecessary variable assignment

* Change to using util library convert

BKPepe committed:
Upgrade youtube_dl to version 2020.01.01 (#30341)

fabaff committed:
Upgrade zeroconf to 0.24.4 (#30347)

frenck committed:
Migrate startca tests from coroutine to async/await (#30354)

frenck committed:
Migrate teksavvy tests from coroutine to async/await (#30353)

frenck committed:
Migrate timer tests from coroutine to async/await (#30352)

frenck committed:
Migrate yr tests from coroutine to async/await (#30351)

JeffLIrion committed:
Fix media_player example for select_source service (#30358)

fabaff committed:
Upgrade python_opendata_transport to 0.2.1 (#30348)

alandtse committed:
Fix HVAC mode for Tesla (#30287)

* Fix HVAC mode for Tesla

* Change HVAC_MODE to HEAT_COOL

AaronDavidSchneider committed:
Add attributes to fritz device_tracker (#30350)

* add attributes to device_tracker

* fix pylint

* requested changes

* docstring updated

frenck committed:
Migrate mailbox tests from coroutine to async/await (#30361)

frenck committed:
Migrate hassio tests from coroutine to async/await (#30363)

frenck committed:
Migrate websocket_api tests from coroutine to async/await (#30364)

frenck committed:
Migrate config tests from coroutine to async/await (#30366)

frenck committed:
Migrate counter tests from coroutine to async/await (#30368)

frenck committed:
Migrate owntracks tests from coroutine to async/await (#30369)

frenck committed:
Migrate no_ip tests from coroutine to async/await (#30370)

frenck committed:
Migrate camera tests from coroutine to async/await (#30372)

frenck committed:
Migrate cast tests from coroutine to async/await (#30374)

frenck committed:
Migrate ffmpeg tests from coroutine to async/await (#30375)

frenck committed:
Migrate discovery tests from coroutine to async/await (#30376)

frenck committed:
Migrate google_* tests from coroutine to async/await (#30377)

alandtse committed:
Add charging rate sensor to Tesla (#30286)

* Add charging rate sensor to Tesla

* Remove reference to bin_type

homeassistant committed:
[ci skip] Translation update

Ernst79 committed:
Restore state for Rfxtrx devices (#30309)

* Restore state rfxtrx switch

* Restore state RFXtrx lights

* Restore state RFXtrx covers

* Restore comment

* Remove line

* Remove logging

* fix black

* Fix typo

Ernst79 committed:
Remove unnessecary rfxtrx light property def (#30397)

scop committed:
Remove unnecessary string literal concatenations (#30360)

frenck committed:
Bump pre-commit to 1.21.0 (#30406)

AaronDavidSchneider committed:
fix issuecomment-570284682 (#30405)

frenck committed:
Migrate local_file tests from coroutine to async/await (#30392)

frenck committed:
Migrate fido tests from coroutine to async/await (#30391)

frenck committed:
Migrate freedns tests from coroutine to async/await (#30390)

frenck committed:
Migrate group tests from coroutine to async/await (#30389)

frenck committed:
Migrate generic tests from coroutine to async/await (#30388)

frenck committed:
Migrate media_player tests from coroutine to async/await (#30387)

frenck committed:
Migrate frontend tests from coroutine to async/await (#30386)

springstan committed:
Bump miflora and bluepy (#30411)

michaeldavie committed:
Bump env_canada to 0.0.31 (#30409)

bendavid committed:
Add generic command functionality to denonavr (#29295)

* Add generic command functionality to denonavr

* add minimal unit tests for denonavr

* fix import order

* simplify denonavr unit test

* handle domain specific service calls with dispatcher

* update unit tests

* update unit tests

* remove unnecessary return value

* fix handling of mock instances in unit tests

bendavid committed:
Add generic command/button functionality to webostv (#30379)

* add generic command/button functionality to webostv

* update codeowners

Kane610 committed:
Axis - Improve tests (#30415)

homeassistant committed:
[ci skip] Translation update

bendavid committed:
Cleanup of state handling in webostv (#30416)

* cleanup unnecessary manipulation of state variables

* update unit test

dmulcahey committed:
Remove ZHA establish device mappings function (#30423)

* remove establish_device_mappings
* inline init

scarface-4711 committed:
Implemented media_play & media_pause / push to version 0.7.11 of denonavr (#30421)

* Implement media_play & media_pause / push to version 0.7.11 of denonavr

* fix denonavr version in requirements_test_all.txt

ochlocracy committed:
Add RTSP stream support for UVC (Unifi Video Client) integration  (#30297)

* Add SUPPORT_STREAM to supported_features.

* Implement stream_source with channel RTSP URIs.

* Add Tests for Stream Support.

* Make stream_source async.

* Removed unused import.

* Re-wrote test to remove warning, and lint error.

tetienne committed:
Fix set tilt position (#30428)

scop committed:
Fix number of times seen in debug message (#30429)

Kane610 committed:
Use config entry unique id for deCONZ (#30122)

* Use config entry unique id

* Clean up

* Backwards compatiblity note

* Fix some of Balloobs comments

* Bump dependency to v66

* Black somehow missed config flow tests...

* Move set unique ID til after possibility to update existing entry

Jc2k committed:
Rework FlowManager to use inheritance (#30133)

* Pull async_finish_flow/async_create_flow out of ConfigEntries

* Towards refactoring

* mypy fixes

* Mark Flow manager with abc.* annotations

* Flake8 fixes

* Mypy fixes

* Blacken data_entry_flow

* Blacken longer signatures caused by mypy changes

* test fixes

* Test fixes

* Fix typo

* Avoid protected member lint (W0212) in config_entries

* More protected member fixes

* Missing await

agners committed:
convert to integer if rounding precision is zero (#30226)

Convert values to integer if rounding precision is zero. With that a
value which is an integer before filtering can be configured to stay
integer when using precision = 0. This also aligns behavior of filters
to how rounding behaves in tempaltes (homeassistant/helpers/template.py,
function forgiving_round).

imduffy15 committed:
Don't allow badly formed upnp devices to kill auto discovery (#30342)

Prevent the following from occurring:

```
2019-12-31 22:38:41 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 85, in async_init
    return await self._async_handle_step(flow, flow.init_step, data)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 145, in _async_handle_step
    result: Dict = await getattr(flow, method)(user_input)
  File "/usr/src/homeassistant/homeassistant/components/deconz/config_flow.py", line 182, in async_step_ssdp
    if discovery_info[ssdp.ATTR_UPNP_MANUFACTURER_URL] != DECONZ_MANUFACTURERURL:
KeyError: 'manufacturerURL'
2019-12-31 22:38:41 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 85, in async_init
    return await self._async_handle_step(flow, flow.init_step, data)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 145, in _async_handle_step
    result: Dict = await getattr(flow, method)(user_input)
  File "/usr/src/homeassistant/homeassistant/components/deconz/config_flow.py", line 182, in async_step_ssdp
    if discovery_info[ssdp.ATTR_UPNP_MANUFACTURER_URL] != DECONZ_MANUFACTURERURL:
KeyError: 'manufacturerURL'
```

elupus committed:
Mark API key as deprecated (#30402)

Service account should be used instead

Kane610 committed:
Axis - Improve tests based on feedback (#30430)

* Use MockConfigEntry, add_to_hass and hass.config_entries.async_setup

* Mock method instead of handling paths

neffs committed:
Update HomeKit position state characteristic for covers (#27867)

* HomeKit: update position state characteristic for covers

position state is a mandatory characteristic for HK window coverings

* Test position state characteristic

sergioisidoro committed:
Handle telegram event commands with args (#30254)

* Handle telegram event commands with args

* Parse message regargless of command

* Lint

* Use multiple assignment

scop committed:
Various string cleanups (#30435)

* Remove some unnecessary string concatenations

* Replace some simple str.formats with f-strings

* Replace some string concatenations with f-strings

basnijholt committed:
Add KEF speakers integration (#28959)

* add KEF speakers platform for the integration

This will work with the KEF LS50 Wireless and KEF LSX speakers.
The development of this code happened on https://github.com/basnijholt/media_player.kef

* rename DATA_KEF -> DOMAIN

* use aiokef v0.2.0 and support LSX and new features

* sort imports

* fix @MartinHjelmare's suggestions

* remove _CONFIGURING

* change STATE_UNKNOWN to None

* use lat and long for unique_id

* bump aiokef to v0.2.2

* use config[ATTR] instead of config.get(ATTR)

* use getmac

* fix case when MAC is None

* use host as instance lifetime id

* fix requirements

Kane610 committed:
WIP deCONZ - Support Fyrtur/Kadrilj battery sensors (#30403)

* Sensor.binary is None means unsupported sensor
* Don't create ordinary sensor on ZHABattery type

Jc2k committed:
isort homekit (#30437)

Jc2k committed:
Don't expose flows that aren't initialised. (#30432)

* Don't expose flows that aren't initialised.

If a flow init does not return immediately then there is a window where our
behaviour is screwy:

 * Can try to configure a flow that isn't ready
 * Can show notifications for discoveries that might yet return an abort

This moves the flow discovery events and notifications to after the flow is
initialised and hides flows that don't have a cur_step from async_progress

* Fix tradfri test

* Black.

* Lint fixes

andersonshatch committed:
Check status field for UPS online binary sensor (#30144)

Avoids issue where status was previously reported as offline when
battery was low or UPS was over voltage

Kane610 committed:
deCONZ - Improve tests based on Martins feedback in Axis integration (#30438)

tchellomello committed:
Bump ring to 0.2.9 (#30439)

Kane610 committed:
Improve tests based on Martins feedback from Axis improve tests (#30442)

Jc2k committed:
Defer homekit_controller initial poll after all entities are created (#30355)

* Make sure first poll happens after sub-platforms are loaded and entities are created.

* Function dosn't need to return anything

* Don't forward entry if already forwarded it

ochlocracy committed:
Implement supported locales for Alexa capabilities (#30285)

* Implement Alexa capability supported locales.

* Check config for supported locales.

* Removed unused import.

* Applied some potpourri to improve code smell.

* Change supported_locales to set.

Kane610 committed:
Axis - Remove manual configuration and legacy config file import (#30365)

* Remove manual configuration and legacy config file import

* Remove unused imports in tests after rebase

dcramer committed:
Add Sentry component (#30422)

balloob committed:
Extract Collection helper from Person integration (#30313)

* Add CRUD foundation

* Use collection helper in person integration

* Lint/pytest

* Add tests

* Lint

* Create notification

brefra committed:
Fix async call inside sync context (#30458)

* Fix async call inside sync context

* Apply black formatting

pnbruckner committed:
Fix counter with empty config (#30454)

elupus committed:
Allow all sources and sound modes to be selected in google assistant (#30418)

* Simplify media player source list

Google don't need a whitelisted set of modes anymore. So let's just forward any mode that we have.

* Report current mode with the setting name, not a synonym

* Refactor mode generation to support other modes

* Support sound mode as mode as well

* Adjust failing test now with sound modes

homeassistant committed:
[ci skip] Translation update

Kane610 committed:
Make Axis integration use config entry unique id (#30461)

* Make Axis integration use config entry unique id

balloob committed:
Only return remote UI url if remote UI enabled (#30453)

balloob committed:
Allow adding unique device ID to mobile app registration (#30457)

* Allow adding unique device ID to mobile app registration

* Remove unknown data

* Better fix

frenck committed:
Bump elgato to 0.2.0 (#30485)

frenck committed:
Fix WLED light transition (#30490)

frenck committed:
Add unique ID to elgato config entries (#30486)

frenck committed:
Bump adguardhome to 0.4.0 (#30483)

frenck committed:
Add unique ID to WLED config entries (#30480)

* Add unique ID to WLED config entries

* Improve robustness of the tests

frenck committed:
Improve WLED pure white support for RGBW (#30477)

frenck committed:
Bump wled to 0.2.1 (#30474)

* Bump wled to 0.2.0

* Bump wled to 0.2.1

frenck committed:
Bump twentemilieu to 0.2.0 (#30471)

Kane610 committed:
deCONZ - Allow manual configuration to update existing configuration (#30469)

* Allow manual configuration to update existing configuration

* Harmonize tests

dmulcahey committed:
Add Zigbee group binding to ZHA (#30433)

* initial group binding work
* add group cluster binding

pnbruckner committed:
Fix timer with empty config (#30463)

balloob committed:
Mobile App add device tracker to person registering app (#30460)

pvizeli committed:
support odroid-n2 on release

pvizeli committed:
bump hassio-builder version

homeassistant committed:
[ci skip] Translation update

Kane610 committed:
Axis - Allow manual configuration to update existing configuration (#30467)

* Allow manual configuration to update existing configuration

* Harmonize tests

* Understand what Elupus means...

balloob committed:
Add person reload service (#30493)

raman325 committed:
Add unique id property for Vizio devices so they get added to entity registry (#30497)

* add unique id property for Vizio devices so they get added to entity registry

* update requirements

* small change to make code more readable

* update to use f-string for consistency

* fix property name

* set unique_id to none if serial number couldnt be found

* set unique_id to none

* ESN should be globally unique so no device type needed. pyvizio will set ESN to None if ESN is unobtainable

balloob committed:
Restore state helper to work with entity registry restoration (#30451)

* Restore state helper to work with entity registry restoratino

* Update restore_state.py

scop committed:
Be consistent with Home Assistant and Hass.io spelling (#30500)

* Be consistent with Home Assistant spelling

* Be consistent with Hass.io spelling

frenck committed:
Migrate automatic tests from coroutine to async/await (#30371)

* Migrate automatic tests from coroutine to async/await

* Remove unneeded side effect

* Replace unittest with asynctest, add additional asserts to tests

balloob committed:
Remove double person data validation (#30501)

Jc2k committed:
Add support for v1 and v2 HomeKit fans. (#30503)

* Add support for v1 and v2 HomeKit fans.

* Lint fix

scop committed:
Preserve new BLE tracker item name if seen before adding discovery (#30318)

* Preserve new BLE device name if one seen before the last scan but not on it

* Save an indented else block

* Use async_fire_time_changed and mock_device_tracker_conf in tests

frenck committed:
Migrate collection of component tests from coroutine to async/await (#30504)

brefra committed:
Add LED control of push buttons and bump velbus-library (#30445)

* Add LED control

* Bump python-velbus library to 2.0.35

To have LED control available in library

* Apply black formating

* Fix no-else-return pylint error

* Changed to f-string and more dry code

* Rewrite turn_on for LED control

homeassistant committed:
[ci skip] Translation update

alandtse committed:
Bump teslajsonpy to 0.2.2 (#30512)

Fixes #30235

Commit:
deCONZ - Rotation support for Xiaomi magic cube (#30472)

* Added rotation to deconz magic cube event

* Device trigger support

scop committed:
Spell Zigbee consistently, with lowercase "b" (#30440)

https://zigbeealliance.org/developer_resources/zigbee-specification/
"Please note that this uses old branding, Zigbee is always a lowercase
 B, not a capital. This change occurred shortly after the official
 release of this document."

Also other docs at https://zigbeealliance.org/solution_type/zigbee/,
https://github.com/zigpy/zigpy/issues/111

GidoHakvoort committed:
Update GarbageTypeCodes for rova

Updating rova/sensor.py. Rova json format changed for garbage collection.

Commit:
Add HmIP-PMFS and HmIP-WHS2 to HomematicIP Cloud (#30325)

* update test data

* Update homematicip_cloud.json

* Add HmIP-WHS2

* Add HmIP-PMFS

benleb committed:
Add surepetcare component (#24426)

* add surepetcare

* cleanup

* remove unused imports and comments

* remove comment

* fix bug which prevented updating the sensors

* improve config validation

* fix voluptuous usage

* fix format & credential storage

* various fixes to hass-conform

* small format fixes

* change False to None

* still trying to be hass-conform

* remove unused class

* fix imports

* fix f-string

* add guard clause?!

* central data fetch

* do not pass in hass, will be provided automatically

* make the linters happy

* disable constant-test warning and add commas

* worksforme

* fix link in manifest

* remove icon

* bump surepy to 0.1.5

* worksforme

* small doc fixes

* add discovery_info guard

* result of another awesome review

* and again :)

* exclude surepetcare in .coveragerc

Jc2k committed:
Add support for homekit air quality sensors (#30510)

* Add air quality sensor

* Fix comment from review

* Fix comment from review

* Lint fix

Jc2k committed:
Set required_features field when registering fan services (#30516)

* Set required_features on fan services

* Fix mqtt fan test

* Update fan.py

bieniu committed:
Add Brother Printer integration (#30359)

* Init entities as unavailable when offline

* Initial commit

* Fix CODEOWNERS

* CODEOWNERS

* Run script.hassfest

* Add initial test

* Bump library

* More tests

* Tests

* Add new sensors and fix KeyError

* Fix unique_id and device_info

* Fix check for configured device

* More tests

* Bump library version

* Add uptime sensor

* Use config entry unique ID

* Run python3 -m script.gen_requirements_all

* Fix pylint error

* Remove pysnmp dependency

* Raise ConfigEntryNotReady when device offline at HA start

* Remove period from logging message

* Generator simplification

* Change raise_on_progress

* Rename data to printer

* Move update state to async_update

* Remove unused _unit_of_measurement

* Remove update of device_info

* Suggested change for tests

* Remove unnecessary argument

* Suggested change

frenck committed:
Format all manifests with prettier (#30521)

balloob committed:
Fix google sync (#30524)

Alexxander0 committed:
Fix AVM FRITZ!DECT switch total consumption (#30478)

actor.get_energy() returns Wh, so division by 1000 is correct

JeffLIrion committed:
Register 'androidtv.download' and 'androidtv.upload' services (#30086)

* Add tests

* Add FileSync test

* Fill in services.yaml for 'androidtv.adb_filesync' service

* Update example paths in services.yaml

* Bump androidtv to 0.0.37

* Bump androidtv to 0.0.37

* Bump androidtv to 0.0.37

* Import LockNotAcquiredException

* Import LockNotAcquiredException from androidtv.exceptions

* Rename 'host' to 'address'

* Add a logging statement when an ADB command is skipped

* Check hass.config.is_allowed_path(local_path)

* Add return

* Fix pylint

* Reduce duplicated code (AndroidTVDevice vs. FireTVDevice)

* Split 'adb_filesync' service into 'download' and 'upload' services

* Don't use '.get()' for required data; return if the services are already registered

* Replace "command" with ATTR_COMMAND

* Don't try to connect to a device if it is a duplicate

rccoleman committed:
Allow specific status codes while notifying mobile_app devices (#30496)

* Removed check and return with a notification status of 201

* Allow additional result codes

Allow 200, 201, and 202 as valid result codes

frenck committed:
Correct integration names in manifests (A-B) (#30527)

* Correct names in manifests (A-B)

* Just Baidu

* Asterisk Call Detail Records

mnigbur committed:
Fire restarted event on restarting active timer (#30475)

* Fire restarted event on restarting active timer

* Added test for the restart event

Quentame committed:
Bump synologydsm to 0.3.0 (#30529)

homeassistant committed:
[ci skip] Translation update

Trekky12 committed:
Add pilight dimmer as light component (#30107)

* Add pilight dimmer as light component

* fix CI errors

* fix missing new lines

* improve formatting and addresses comments of @springstan

* rename config parameter and remove super() call to match pylint

* import only used constants of the pilight component

* Add myself to the code owners

* fix CODEOWNERS

Commit:
Add "autobypass" option when arming AlarmDecoder integration (#30002)

* Initial implementation

* Passing autobypass parameter to constructor, as suggested

* Black formatting

* Removed default value from autobypass parameter of the constructor, as it's redundant

raman325 committed:
Update vizio component to prepare for config flow and zeroconf support (#30522)

* move constants, move schema validation to init, prepare for zeroconf/config flow

* update requirements_all and make setup validation more pythonic

* mistake in refactor

* un-revert change from previous PR

* rename schema var

* update .coveragerc, move validate_auth to __init__, only attempt to get ESN if device setup is validated

* reinstate validate_setup but change logic to use new connection test function

* fix schema validation

* simplify string empty check logic

* missed commit

frenck committed:
Correct names in manifests (C-L) (#30532)

Adminiuga committed:
Use collection helpers for input_boolean (#30514)

* Refactor input_boolean to use config dict for instantiation.

* Refactor input_boolean to use YamlCollection.

* Add storage collection to input_boolean.

* Update homeassistant/components/input_boolean/__init__.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

* Doh.

* Address comments.

* Add editable device state attribute.

* Clean up entities from entity registry on removal.

Reload yaml from correct source.

* Add tests.

* Update homeassistant/components/input_boolean/__init__.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

* Don't reset entity state on updates.

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

csfreak committed:
Add venstar PIN to config and client initialization (#30300)

* Add PIN to config and client  initialization

* Fix Formatting Error

* update PLATFORM_SCHEMA validation with the new option.

colinfrei committed:
Swiss Transport Data: Save delay on sensor (#30526)

casperklein committed:
Update manifest.json: Typo fixed (#30542)

balloob committed:
Upgrade to use new version of HASS-nabucasa (#30506)

* Upgrade to use new version of HASS-nabucasa

* Update hass-nabucasa to 0.31

frenck committed:
Correct names in manifests (M-P) (#30541)

JeffLIrion committed:
Bump adb-shell to 0.1.1 and androidtv to 0.0.38 (#30539)

* Bump adb-shell to 0.1.1 and androidtv to 0.0.38

* Bump adb-shell to 0.1.1 and androidtv to 0.0.38

* Bump adb-shell to 0.1.1 and androidtv to 0.0.38

JeffLIrion committed:
Catch UnicodeDecodeError exceptions in 'androidtv.adb_command' service (#30538)

* Catch UnicodeDecodeError exceptions in 'androidtv.adb_command' service

* Replace "adb_command" with SERVICE_ADB_COMMAND

raman325 committed:
Remove `suppress_warning` from config options for vizio component (#30536)

* move constants, move schema validation to init, prepare for zeroconf/config flow

* mistake in refactor

* update .coveragerc, move validate_auth to __init__, only attempt to get ESN if device setup is validated

* fix schema validation

* remove suppress warning config option

* remove config_flow from manifest since its not ready yet

* update requirements_all.txt

* fix issue with merge conflict

balloob committed:
Use capability attributes in climate (#30544)

Co-authored-by: Jc2k <john.carr@unrouted.co.uk>

frenck committed:
Correct names in manifests (Q-S) (#30543)

springstan committed:
Update featured integrations screenshot in README (#30134)

* Update featured integrations in README

* Update screenshot to include Somfy and exclude Almond

* Update outdated screenshot of featured components

frenck committed:
Correct names in manifests (T-Z) (#30546)

alemuro committed:
Add TMB sensor (#27964)

* Add TMB i-Bus sensor

* Raise exception if HTTP error

* Addressed comments

* Fixed format

* Fixed format

* Addressed comments

* Remove guard before add_entities method

* Remove unauthorized exception

* Make CONF_BUS_STOPS option required

Commit:
Add reset_energy_counter service to Homematic IP Cloud (#30256)

* Add reset_energy_counter service to Homematic IP Cloud

* Fix isort

* Register service as admin service

frenck committed:
Add Integration Quality Scale to manifest (#30547)

tetienne committed:
Ease code before adding color and temperature to light template (#30455)

* Split async_update

* Use pytest parameters to avoid duplicate code

* Fix UnboundLocalError

* Test error rendering for template state

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

balloob committed:
Entity Component to no longer generate automatic groups (#23789)

bachya committed:
Fix incorrect Rainmachine dict lookups (#30550)

Commit:
Fix missing strings for deconz magic cube rotation support (#30552)

* Fix missing strings for deconz magic cube rotation

* Fix missing strings for deconz magic cube rotation

InuSasha committed:
Fix hue flashing innr lights (#29828)

frenck committed:
Migrate Integration Quality Scale from docs to manifest: internal (#30551)

bramkragten committed:
Updated frontend to 20200107.0 (#30561)

bramkragten committed:
Add command to delete lovelace config (#30558)

* Add command to delete lovelace config

* Test + decorator for WS

homeassistant committed:
[ci skip] Translation update

dshokouhi committed:
Add android repo to issue template (#30571)

Ernst79 committed:
Bump pyRFXtrx to 0.25 (#30566)

Kane610 committed:
deCONZ - Make sensors sorted on creation for deterministic order (#30569)

* Make sensors sorted on creation for deterministic order

* Martins comment

Kane610 committed:
deCONZ - Fix device automations (#30577)

Santobert committed:
Trigger automation without skipping condition (#28484)

* Trigger automation without skipping condition

* [] instead of .get()

* Update __init__.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

frenck commented:
@dlashua Small and friendly ping, you are close to the finish, there is one small comment to address and a merge conflict, could you take a look? 👍

Adminiuga committed:
Use more consts for person implementation (#30553)

* Use more consts for person implementation.

* Use ATTR_ENTITY_ID.

dlashua commented:
@frenck What do I do with @MartinHjelmare 's comment? It's not incorrect. But neither is what I have. My version uses two if statements but makes the comment that @balloob asked for more clear. And @MartinHjelmare 's suggestion would use one if statement. The comment was left after @balloob approved the PR, so I assumed it wasn't stopping the PR and I liked it better the way I wrote it, which is why I wrote it that way.

And I don't know when the conflict showed up, but it was sometime after that.

dlashua committed:
use an f-string

MartinHjelmare commented:
Why is my comment not correct?

dlashua commented:
@frenck The conflict is because some other PR converted a piece of this to use an f-string instead of concating a string. I changed my code to use the f-string also, but the conflict isn't resolving. I think because the f-string line of code is too close to other lines of code that I changed.

frenck committed:
Migrate Integration Quality Scale from docs to manifest (#30579)

dlashua committed:
change style again

dlashua committed:
change style again

dlashua committed:
add space to fix conflict

dlashua committed:
add space to fix conflict

dlashua committed:
add more space to fix conflict

dlashua committed:
add more space to fix conflict

MartinHjelmare reviewed (approved):
Thanks!

dlashua committed:
delete spaces to resolve conflict

MartinHjelmare commented:
I think you need to rebase to solve the merge conflict.

MartinHjelmare commented:
Nevermind, it's good now.

frenck commented:
For more information on rebasing/catching up with reality see:

<https://developers.home-assistant.io/docs/en/development_catching_up.html>

dlashua commented:
All conflicts are resolved. And @MartinHjelmare 's suggestion has been included.

MartinHjelmare commented:
Looks like flake8 run failed. Please address the flake8 warnings in the build.

```
homeassistant/components/netgear/device_tracker.py:113:1: W293 blank line contains whitespace
homeassistant/components/netgear/device_tracker.py:126:1: W293 blank line contains whitespace
homeassistant/components/netgear/device_tracker.py:127:5: E303 too many blank lines (2)
```

dlashua committed:
Merge branch 'patch-1' of github.com:dlashua/home-assistant into patch-1

dlashua commented:
This PR is now totally messed up. I'll close it.

Quentame commented:
Please rebase, not merge.

It won't request everyone who code between :wink:

dlashua commented:
Please see PR #30581.
